trestle/core/control_io.py

```diff
@@ -807,71 +737,236 @@ def _read_sections(ii: int, lines: List[str], control_id: str,
+            ControlIOReader._add_node_to_dict(comp_name, label, comp_dict, subnode)
+
     @staticmethod
     def read_all_implementation_prose_and_header(
         control_file: pathlib.Path
-    ) -> Tuple[Dict[str, List[str]], Dict[str, List[str]]]:
+    ) -> Tuple[Dict[str, Dict[str, List[str]]], Dict[str, List[str]]]:
         """
         Find all labels and associated prose in this control.
 
         Args:
             control_file: path to the control markdown file
 
         Returns:
-            Dictionary of part labels and corresponding prose read from the markdown file.
+            Dictionary by comp_name of Dictionaries of part labels and corresponding prose read from the markdown file.
+            Also returns the yaml header as dict in second part of tuple.
+            This does not generate components - it only tracks component names and associated responses.
         """
+        comp_dict = {}
+        yaml_header = {}
+        # this level only adds for top level component but add_node_to_dict can add for other components
+        comp_name = const.SSP_MAIN_COMP_NAME
         try:
             if not control_file.exists():
-                return {}, {}
+                return comp_dict, yaml_header
             md_api = MarkdownAPI()
-            header, _ = md_api.processor.process_markdown(control_file)
-
-            lines = ControlIOReader._load_control_lines(control_file)
-            ii = 0
-            # keep moving down through the file picking up labels and prose
-            responses: Dict[str, List[str]] = {}
-            while True:
-                ii, part_label, prose_lines = ControlIOReader._read_label_prose(ii, lines)
-                while prose_lines and not prose_lines[0].strip(' \r\n'):
-                    del prose_lines[0]
-                while prose_lines and not prose_lines[-1].strip(' \r\n'):
-                    del prose_lines[-1]
-                if part_label and prose_lines:
-                    responses[part_label] = prose_lines
-                if ii < 0:
-                    break
+            yaml_header, control = md_api.processor.process_markdown(control_file)
+
+            imp_string = 'Implementation'
+            headers = control.get_all_headers_for_key(imp_string, False)
+            if not headers:
+                # if statement has no parts there is only one response for entire control
+                headers = control.get_all_headers_for_key(const.SSP_MD_IMPLEMENTATION_QUESTION, False)
+                # should be only one header
+                for header in headers:
+                    node = control.get_node_for_key(header)
+                    ControlIOReader._add_node_to_dict(comp_name, 'Statement', comp_dict, node)
+            else:
+                for header in headers:
+                    tokens = header.split(' ', 2)
+                    if tokens[0] == '##' and tokens[1] == imp_string:
+                        label = tokens[2].strip()
```

Comment:
As in the comment above, probably worth cleaning header in separate function. Then this could be simplified to two calls: `MarkdownNode.get_clean_key()` and `MarkdownNode.get_node_header_lvl()`